libyul/optimiser/LoadResolver.cpp

```diff
@@ -71,8 +102,64 @@ void LoadResolver::tryResolve(
 			if (inScope(*value))
 				_e = Identifier{locationOf(_e), *value};
 	}
-	else if (m_optimizeMLoad && _location == StoreLoadLocation::Memory)
+	else if (!m_containsMSize && _location == StoreLoadLocation::Memory)
 		if (auto value = util::valueOrNullptr(m_memory, key))
 			if (inScope(*value))
 				_e = Identifier{locationOf(_e), *value};
 }
+
+void LoadResolver::tryEvaluateKeccak(
+	Expression& _e,
+	std::vector<Expression> const& _arguments
+)
+{
+	// The costs are only correct for hashes of 32 bytes or 1 word (when rounded up).
+	GasMeter gasMeter{
+		dynamic_cast<EVMDialect const&>(m_dialect),
+		!m_expectedExecutionsPerDeployment,
+		m_expectedExecutionsPerDeployment ? *m_expectedExecutionsPerDeployment : 1
+	};
+
+	auto costOfKeccak = gasMeter.costs(_e);
+	auto costOfLiteral = gasMeter.costs(
+		Literal{
+			{},
+			LiteralKind::Number,
+			// a dummy 256-bit number to represent the Keccak256 hash.
+			YulString{numeric_limits<u256>::max().str()},
+			{}
+		}
+	);
+
+	// We skip if there are no net gas savings.
+	// Note that for default `m_runs = 200`, the values are
+	// `costOfLiteral = 7200` and `costOfKeccak = 9000` for runtime context.
+	// For creation context: `costOfLiteral = 531` and `costOfKeccak = 90`.
+	if (costOfLiteral > costOfKeccak)
+		return;
+
+	yulAssert(_arguments.size() == 2, "");
+	Identifier const* memoryKey = std::get_if<Identifier>(&_arguments.at(0));
+	Identifier const* length = std::get_if<Identifier>(&_arguments.at(1));
+
+	if (!memoryKey || !length)
+		return;
+
+	auto memoryValue = util::valueOrNullptr(m_memory, memoryKey->name);
+	if (memoryValue && inScope(*memoryValue))
```

Comment:
Do you need the `inScope` check?

Reply:
I think it's redundant, but it was added during `DataFlowAnalyzer` refactoring (https://github.com/ethereum/solidity/pull/10467/commits/75bed10be56939d7b4a28bfc9fa006f4fe1267d9) to make `LoadResolver` slightly safer.

Reply:
Ok, but it's not relevant here, because we don't use it directly, is it?

Reply:
We still get the value to hash from `memoryValue`. So I think the check makes it slightly safer, although redundant.